File: Workspace/TwentyFortyEightDemo/src/TwentyFortyEight.java
import java.util.ArrayList;
import java.util.Random;
public class TwentyFortyEight extends ProblemState
{
  // The grid that contains the game board
  private int[][] grid =new int[4][4];
  // The direction of the last move
  private String  direction;
  // Initial state constructor
  public TwentyFortyEight()
  {
    super();
    int i,j;
    // set the pieces;
    for(i=0;i<4;i++)
    {
      for(j=0;j<4;j++)
      {
        // -1 is blank
        grid[i][j]=-1;
      }
    }
    // Add a random tile to the board in a blank spot
    addRandom();
  }
  // Derived state constructor
  public TwentyFortyEight(TwentyFortyEight orig)
  {
    super(orig);
    setCost(orig.getCost()+1);
    int i,j;
    // Copy all of the pieces (don't add a random yet)
    for(i=0;i<4;i++)
    {
      for(j=0;j<4;j++)
      {
        grid[i][j]=orig.grid[i][j];
      }
    }
  }
  private void addRandom()
  {
    // Add either a 2 or a 4 to a random blank spot on the board
    Random rand=new Random();
    int nextNum=0;
    int randomNum=rand.nextInt(2);
    if(randomNum==0)
    {
      nextNum=2;
    }else
    {
      nextNum=4;
    }
    // find the potential places to place the new tile
    int i,j;
    ArrayList<Integer> potentialPlaces=new ArrayList<Integer>();
    for(i=0;i<4;i++)
    {
      for(j=0;j<4;j++)
      {
        if(grid[i][j]==-1)
        {
          // Add a potential place to put the new tile
          potentialPlaces.add(i*4+j);
        }
      }
    }
    if(potentialPlaces.size()==0){ return; }
    randomNum=rand.nextInt(potentialPlaces.size());
    int location=potentialPlaces.get(randomNum);
    // Place the new tile in a random but acceptable location (don't overwrite a
    // non-blank cell)
    grid[location/4][location%4]=nextNum;
  }
  // Get all of the moves that we can go to from this state
  @Override public ArrayList<ProblemState> createSuccessorStates()
  {
    ArrayList<ProblemState> states=new ArrayList<ProblemState>(4);
    // Swipe left
    TwentyFortyEight left=new TwentyFortyEight(this);
    int i;
    // Check each row left to right looking for combos and packing things
    // together
    for(i=0;i<4;i++)
    {
      int[] array=new int[4];
      array[0]=grid[i][0];
      array[1]=grid[i][1];
      array[2]=grid[i][2];
      array[3]=grid[i][3];
      // Pack the tiles together then look for tiles to combine
      array=packAndCombine(array);
      left.grid[i][0]=array[0];
      left.grid[i][1]=array[1];
      left.grid[i][2]=array[2];
      left.grid[i][3]=array[3];
    }
    left.direction="left";
    // Ensure that the new state gets us somewhere useful before adding it
    if(!equals(left))
    {
      // Add a random tile
      left.addRandom();
      states.add(left);
    }
    // Swipe right
    TwentyFortyEight right=new TwentyFortyEight(this);
    // Check each row right to left looking for combos and packing things
    // together
    for(i=0;i<4;i++)
    {
      int[] array=new int[4];
      array[0]=grid[i][3];
      array[1]=grid[i][2];
      array[2]=grid[i][1];
      array[3]=grid[i][0];
      // Pack the tiles together then look for tiles to combine
      array=packAndCombine(array);
      right.grid[i][3]=array[0];
      right.grid[i][2]=array[1];
      right.grid[i][1]=array[2];
      right.grid[i][0]=array[3];
    }
    right.direction="right";
    // Ensure that the new state gets us somewhere useful before adding it
    if(!equals(right))
    {
      right.addRandom();
      states.add(right);
    }
    // Swipe up
    TwentyFortyEight up=new TwentyFortyEight(this);
    // Check each column up to down looking for combos and packing things
    // together
    for(i=0;i<4;i++)
    {
      int[] array=new int[4];
      array[0]=grid[0][i];
      array[1]=grid[1][i];
      array[2]=grid[2][i];
      array[3]=grid[3][i];
      // Pack the tiles together then look for tiles to combine
      array=packAndCombine(array);
      up.grid[0][i]=array[0];
      up.grid[1][i]=array[1];
      up.grid[2][i]=array[2];
      up.grid[3][i]=array[3];
    }
    up.direction="up";
    // Ensure that the new state gets us somewhere useful before adding it
    if(!equals(up))
    {
      up.addRandom();
      states.add(up);
    }
    // Swipe down
    TwentyFortyEight down=new TwentyFortyEight(this);
    // Check each column down to up looking for combos and packing things
    // together
    for(i=0;i<4;i++)
    {
      int[] array=new int[4];
      array[0]=grid[3][i];
      array[1]=grid[2][i];
      array[2]=grid[1][i];
      array[3]=grid[0][i];
      // Pack the tiles together then look for tiles to combine
      array=packAndCombine(array);
      down.grid[3][i]=array[0];
      down.grid[2][i]=array[1];
      down.grid[1][i]=array[2];
      down.grid[0][i]=array[3];
    }
    down.direction="down";
    // Ensure that the new state gets us somewhere useful before adding it
    if(!equals(down))
    {
      down.addRandom();
      states.add(down);
    }
    return states;
  }
  // Overridden equals method for the board state
  @Override public boolean equals(Object obj)
  {
    if(obj==null){ return false; }
    if(getClass()!=obj.getClass()){ return false; }
    final TwentyFortyEight that=(TwentyFortyEight)obj;
    int i,j;
    for(i=0;i<4;i++)
    {
      for(j=0;j<4;j++)
      {
        // If one tile is not the same, they are not equal
        if(grid[i][j]!=that.grid[i][j]){ return false; }
      }
    }
    return true;
  }
  // Have we reached the goal?
  @Override public boolean goalTest()
  {
    int i,j;
    for(i=0;i<4;i++)
    {
      for(j=0;j<4;j++)
      {
        // We are going to 2048 (changed to 4096, 8192, etc for more difficult
        // problem
        if(grid[i][j]==2048){ return true; }
      }
    }
    return false;
  }
  @Override public boolean isLegal()
  {
    // Not useful in this case
    return true;
  }
  // Pack tiles in the direction that we are swiping and combine them
  private int[] packAndCombine(int[] array)
  {
    int j=0;
    while(j<4)
    {
      // Start at one side of the row/column looking for a blank then shifting
      // following tiles towards it
      if(array[j]==-1)
      {
        // Save the iterator
        int savedJ=j;
        // Look for the next non-blank
        while(j<4)
        {
          if(array[j]==-1)
          {
            j++;
          }else
          {
            // Move the non-blank tile down
            array[savedJ]=array[j];
            array[j]=-1;
            break;
          }
        }
        // Restore the iterator
        j=savedJ;
      }
      // Try to combine (tiles must not be blank, must be equal, and must be
      // adjacent
      // if(j!=3&&array[j]!=-1&&array[j]==array[j+1])
      // {
      // Multiply the original tile by two
      // array[j]*=2;
      // Clear the other tile
      // array[j+1]=-1;
      // }
      // Move to the next file (we can now ignore the one we just came from)
      j++;
    }
    for(j=0;j<3;j++)
    {
      if(array[j]!=-1&&array[j]==array[j+1])
      {
        array[j]*=2;
        for(int k=j+1;k<3;k++)
        {
          array[k]=array[k+1];
        }
        array[3]=-1;
      }
    }
    return array;
  }
  // Print the state in a readable format
  @Override public void printState()
  {
    System.out.println("+------+------+------+------+  "+direction);
    int i,j;
    for(i=0;i<4;i++)
    {
      System.out.println("|      |      |      |      |");
      System.out.print("|");
      for(j=0;j<4;j++)
      {
        if(grid[i][j]==-1)
        {
          System.out.print("      |");
        }else
        {
          // Format to four digits in the printed result
          System.out.printf(" %4d |",grid[i][j]);
        }
      }
      System.out.println();
      System.out.println("|      |      |      |      |");
      System.out.println("+------+------+------+------+");
    }
    System.out.println("Moves made = "+getCost());
  }
}

File: Projects/TwentyFortyEight/twentyfortyeight.cpp
#include "twentyfortyeight.h"
#include<stdlib.h>
#include <time.h>
using namespace std;

TwentyFortyEight::TwentyFortyEight()
{
  this->cost=0;
  this->grid[0]=0;
  this->grid[1]=0;
  this->grid[2]=0;
  this->grid[3]=0;
  this->grid[4]=0;
  this->grid[5]=0;
  this->grid[6]=0;
  this->grid[7]=0;
  this->grid[8]=0;
  this->grid[9]=0;
  this->grid[10]=0;
  this->grid[11]=0;
  this->grid[12]=0;
  this->grid[13]=0;
  this->grid[14]=0;
  this->grid[15]=0;
  this->addRandom();
}
TwentyFortyEight::TwentyFortyEight(ProblemState*orig)
{
  TwentyFortyEight*original=(TwentyFortyEight*)orig;
  this->setCost(original->getCost()+1);
  this->grid[0]=original->grid[0];
  this->grid[1]=original->grid[1];
  this->grid[2]=original->grid[2];
  this->grid[3]=original->grid[3];
  this->grid[4]=original->grid[4];
  this->grid[5]=original->grid[5];
  this->grid[6]=original->grid[6];
  this->grid[7]=original->grid[7];
  this->grid[8]=original->grid[8];
  this->grid[9]=original->grid[9];
  this->grid[10]=original->grid[10];
  this->grid[11]=original->grid[11];
  this->grid[12]=original->grid[12];
  this->grid[13]=original->grid[13];
  this->grid[14]=original->grid[14];
  this->grid[15]=original->grid[15];
  this->direction=-1;
}
TwentyFortyEight::~TwentyFortyEight()
{
  //Nothing to do
}
TwentyFortyEight TwentyFortyEight::operator=(const TwentyFortyEight& that)
{
  if (this == &that)
  {
    return*this;
  }
  this->grid[0]=that.grid[0];
  this->grid[1]=that.grid[1];
  this->grid[2]=that.grid[2];
  this->grid[3]=that.grid[3];
  this->grid[4]=that.grid[4];
  this->grid[5]=that.grid[5];
  this->grid[6]=that.grid[6];
  this->grid[7]=that.grid[7];
  this->grid[8]=that.grid[8];
  this->grid[9]=that.grid[9];
  this->grid[10]=that.grid[10];
  this->grid[11]=that.grid[11];
  this->grid[12]=that.grid[12];
  this->grid[13]=that.grid[13];
  this->grid[14]=that.grid[14];
  this->grid[15]=that.grid[15];
  this->cost=that.cost;
  this->direction=that.direction;
  return*this;
}
bool TwentyFortyEight::operator==(const TwentyFortyEight&that) const
{
  if(this==&that)
  {
    return true;
  }
  if(this->grid[0]!=that.grid[0])return false;
  if(this->grid[1]!=that.grid[1])return false;
  if(this->grid[2]!=that.grid[2])return false;
  if(this->grid[3]!=that.grid[3])return false;
  if(this->grid[4]!=that.grid[4])return false;
  if(this->grid[5]!=that.grid[5])return false;
  if(this->grid[6]!=that.grid[6])return false;
  if(this->grid[7]!=that.grid[7])return false;
  if(this->grid[8]!=that.grid[8])return false;
  if(this->grid[9]!=that.grid[9])return false;
  if(this->grid[10]!=that.grid[10])return false;
  if(this->grid[11]!=that.grid[11])return false;
  if(this->grid[12]!=that.grid[12])return false;
  if(this->grid[13]!=that.grid[13])return false;
  if(this->grid[14]!=that.grid[14])return false;
  if(this->grid[15]!=that.grid[15])return false;
  return true;
}
bool TwentyFortyEight::operator !=(const TwentyFortyEight&that) const
{
  return !(*this==that);
}
vector<shared_ptr<ProblemState> > TwentyFortyEight::createSuccessorStates()
{
  vector<shared_ptr<ProblemState> > states;
  states.reserve(4);
  TwentyFortyEight left = new TwentyFortyEight(this);
  unsigned short int* array = new unsigned short int[4];
  array[0]=this->grid[0];
  array[1]=this->grid[1];
  array[2]=this->grid[2];
  array[3]=this->grid[3];
  array=packAndCombine(array);
  left.grid[0]=array[0];
  left.grid[1]=array[1];
  left.grid[2]=array[2];
  left.grid[3]=array[3];
  array[0]=this->grid[4];
  array[1]=this->grid[5];
  array[2]=this->grid[6];
  array[3]=this->grid[7];
  array=packAndCombine(array);
  left.grid[4]=array[0];
  left.grid[5]=array[1];
  left.grid[6]=array[2];
  left.grid[7]=array[3];
  array[0]=this->grid[8];
  array[1]=this->grid[9];
  array[2]=this->grid[10];
  array[3]=this->grid[11];
  array=packAndCombine(array);
  left.grid[8]=array[0];
  left.grid[9]=array[1];
  left.grid[10]=array[2];
  left.grid[11]=array[3];
  array[0]=this->grid[12];
  array[1]=this->grid[13];
  array[2]=this->grid[14];
  array[3]=this->grid[15];
  array=packAndCombine(array);
  left.grid[12]=array[0];
  left.grid[13]=array[1];
  left.grid[14]=array[2];
  left.grid[15]=array[3];
  left.direction=0;
  if(*this!=left)
  {
    left.addRandom();
    states.push_back(shared_ptr<TwentyFortyEight>(&left));
  }

  TwentyFortyEight right = new TwentyFortyEight(this);
  array[0]=this->grid[3];
  array[1]=this->grid[2];
  array[2]=this->grid[1];
  array[3]=this->grid[0];
  array=packAndCombine(array);
  right.grid[3]=array[0];
  right.grid[2]=array[1];
  right.grid[1]=array[2];
  right.grid[0]=array[3];
  array[0]=this->grid[7];
  array[1]=this->grid[6];
  array[2]=this->grid[5];
  array[3]=this->grid[4];
  array=packAndCombine(array);
  right.grid[7]=array[0];
  right.grid[6]=array[1];
  right.grid[5]=array[2];
  right.grid[4]=array[3];
  array[0]=this->grid[11];
  array[1]=this->grid[10];
  array[2]=this->grid[9];
  array[3]=this->grid[8];
  array=packAndCombine(array);
  right.grid[11]=array[0];
  right.grid[10]=array[1];
  right.grid[9]=array[2];
  right.grid[8]=array[3];
  array[0]=this->grid[15];
  array[1]=this->grid[14];
  array[2]=this->grid[13];
  array[3]=this->grid[12];
  array=packAndCombine(array);
  right.grid[15]=array[0];
  right.grid[14]=array[1];
  right.grid[13]=array[2];
  right.grid[12]=array[3];
  right.direction=1;
  if(*this!=right)
  {
    right.addRandom();
    states.push_back(shared_ptr<TwentyFortyEight>(&right));
  }

  TwentyFortyEight up = new TwentyFortyEight(this);

  array[0]=this->grid[0];
  array[1]=this->grid[4];
  array[2]=this->grid[8];
  array[3]=this->grid[12];
  array=packAndCombine(array);
  up.grid[0]=array[0];
  up.grid[4]=array[1];
  up.grid[8]=array[2];
  up.grid[12]=array[3];
  array[0]=this->grid[1];
  array[1]=this->grid[5];
  array[2]=this->grid[9];
  array[3]=this->grid[13];
  array=packAndCombine(array);
  up.grid[1]=array[0];
  up.grid[5]=array[1];
  up.grid[9]=array[2];
  up.grid[13]=array[3];
  array[0]=this->grid[2];
  array[1]=this->grid[6];
  array[2]=this->grid[10];
  array[3]=this->grid[14];
  array=packAndCombine(array);
  up.grid[2]=array[0];
  up.grid[6]=array[1];
  up.grid[10]=array[2];
  up.grid[14]=array[3];
  array[0]=this->grid[3];
  array[1]=this->grid[7];
  array[2]=this->grid[11];
  array[3]=this->grid[15];
  array=packAndCombine(array);
  up.grid[3]=array[0];
  up.grid[7]=array[1];
  up.grid[11]=array[2];
  up.grid[15]=array[3];
  up.direction=2;
  if(*this!=up)
  {
    up.addRandom();
    states.push_back(shared_ptr<TwentyFortyEight>(&up));
  }

  TwentyFortyEight down = new TwentyFortyEight(this);
  array[0]=this->grid[12];
  array[1]=this->grid[8];
  array[2]=this->grid[4];
  array[3]=this->grid[0];
  array=packAndCombine(array);
  down.grid[12]=array[0];
  down.grid[8]=array[1];
  down.grid[4]=array[2];
  down.grid[0]=array[3];
  array[0]=this->grid[13];
  array[1]=this->grid[9];
  array[2]=this->grid[5];
  array[3]=this->grid[1];
  array=packAndCombine(array);
  down.grid[13]=array[0];
  down.grid[9]=array[1];
  down.grid[4]=array[2];
  down.grid[1]=array[3];
  array[0]=this->grid[14];
  array[1]=this->grid[10];
  array[2]=this->grid[6];
  array[3]=this->grid[2];
  array=packAndCombine(array);
  down.grid[14]=array[0];
  down.grid[10]=array[1];
  down.grid[6]=array[2];
  down.grid[2]=array[3];
  array[0]=this->grid[15];
  array[1]=this->grid[11];
  array[2]=this->grid[7];
  array[3]=this->grid[3];
  array=packAndCombine(array);
  down.grid[15]=array[0];
  down.grid[11]=array[1];
  down.grid[7]=array[2];
  down.grid[3]=array[3];
  down.direction=3;
  if(*this!=down)
  {
    down.addRandom();
    states.push_back(shared_ptr<TwentyFortyEight>(&down));
  }
  delete [] array;
  return states;
}

bool TwentyFortyEight::goalTest()
{
  if(grid[0]==goal)return true;
  if(grid[1]==goal)return true;
  if(grid[2]==goal)return true;
  if(grid[3]==goal)return true;
  if(grid[4]==goal)return true;
  if(grid[5]==goal)return true;
  if(grid[6]==goal)return true;
  if(grid[7]==goal)return true;
  if(grid[8]==goal)return true;
  if(grid[9]==goal)return true;
  if(grid[10]==goal)return true;
  if(grid[11]==goal)return true;
  if(grid[12]==goal)return true;
  if(grid[13]==goal)return true;
  if(grid[14]==goal)return true;
  if(grid[15]==goal)return true;
  return false;
}

void TwentyFortyEight::printState()
{
  printf("+------+------+------+------+ $s\n", this->direction);
  printf("|      |      |      |      |\n|");
  if(!this->grid[0])printf("      |");
  else printf(" %4d |",this->grid[0]);
  if(!this->grid[1])printf("      |");
  else printf(" %4d |",this->grid[1]);
  if(!this->grid[2])printf("      |");
  else printf(" %4d |",this->grid[2]);
  if(!this->grid[3])printf("      |");
  else printf(" %4d |",this->grid[3]);
  printf("\n|      |      |      |      |\n+------+------+------+------+\n");
  printf("|      |      |      |      |\n|");
  if(!this->grid[4])printf("      |");
  else printf(" %4d |",this->grid[4]);
  if(!this->grid[5])printf("      |");
  else printf(" %4d |",this->grid[5]);
  if(!this->grid[6])printf("      |");
  else printf(" %4d |",this->grid[6]);
  if(!this->grid[7])printf("      |");
  else printf(" %4d |",this->grid[7]);
  printf("\n|      |      |      |      |\n+------+------+------+------+\n");
  printf("|      |      |      |      |\n|");
  if(!this->grid[8])printf("      |");
  else printf(" %4d |",this->grid[8]);
  if(!this->grid[9])printf("      |");
  else printf(" %4d |",this->grid[9]);
  if(!this->grid[10])printf("      |");
  else printf(" %4d |",this->grid[10]);
  if(!this->grid[11])printf("      |");
  else printf(" %4d |",this->grid[11]);
  printf("\n|      |      |      |      |\n+------+------+------+------+\n");
  printf("|      |      |      |      |\n|");
  if(!this->grid[12])printf("      |");
  else printf(" %4d |",this->grid[12]);
  if(!this->grid[13])printf("      |");
  else printf(" %4d |",this->grid[13]);
  if(!this->grid[14])printf("      |");
  else printf(" %4d |",this->grid[14]);
  if(!this->grid[15])printf("      |");
  else printf(" %4d |",this->grid[15]);
  printf("\n|      |      |      |      |\n+------+------+------+------+\n");
  printf("Moves made = %d\n\n", this->cost);
}

void TwentyFortyEight::addRandom()
{
  srand((unsigned int)time(NULL));
  char nextNum;
  if(rand()%1+1)
  {
    nextNum=2;
  }
  else
  {
    nextNum=4;
  }
  vector<char> potentialPlaces;
  if(!this->grid[0])potentialPlaces.push_back(0);
  if(!this->grid[1])potentialPlaces.push_back(1);
  if(!this->grid[2])potentialPlaces.push_back(2);
  if(!this->grid[3])potentialPlaces.push_back(3);
  if(!this->grid[4])potentialPlaces.push_back(4);
  if(!this->grid[5])potentialPlaces.push_back(5);
  if(!this->grid[6])potentialPlaces.push_back(6);
  if(!this->grid[7])potentialPlaces.push_back(7);
  if(!this->grid[8])potentialPlaces.push_back(8);
  if(!this->grid[9])potentialPlaces.push_back(9);
  if(!this->grid[10])potentialPlaces.push_back(10);
  if(!this->grid[11])potentialPlaces.push_back(11);
  if(!this->grid[12])potentialPlaces.push_back(12);
  if(!this->grid[13])potentialPlaces.push_back(13);
  if(!this->grid[14])potentialPlaces.push_back(14);
  if(!this->grid[15])potentialPlaces.push_back(15);
  if(!potentialPlaces.size())return;
  char randomNum=rand()%potentialPlaces.size();
  this->grid[potentialPlaces[randomNum]]=nextNum;
}
unsigned short int* TwentyFortyEight::packAndCombine(unsigned short int *array)
{
  char i=0,j;
  while(i<4)
  {
    if(!array[i])
    {
      j=i;
      while(i<4)
      {
        if(!array[i])
        {
          i++;
        }
        else
        {
          array[j]=array[i];
          array[i]=0;
          break;
        }
      }
      i=j;
    }
    if(i!=3&&array[i]&&array[i]==array[i+1])
    {
      array[i]*=2;
      array[i+1]=0;
    }
    i++;
  }
  return array;
}


